lib/spack/spack/verify.py

```diff
@@ -0,0 +1,250 @@
+            if sys.version_info[0] >= 3:
+                js = js.encode()
+            f.write(js)
+
+        fp.set_permissions_by_spec(manifest_file, spec)
+
+
+def check_entry(path, data):
+    res = VerificationResults()
+
+    if not data:
+        res.add_error(path, 'added')
+        return res
+
+    stat = os.stat(path)
+
+    # Check for all entries
+    if stat.st_mode != data['mode']:
+        res.add_error(path, 'mode')
+    if stat.st_uid != data['owner']:
+        res.add_error(path, 'owner')
+    if stat.st_gid != data['group']:
+        res.add_error(path, 'group')
+
+    # Check for symlink targets  and listed as symlink
+    if os.path.islink(path):
+        if data['type'] != 'link':
+            res.add_error(path, 'type')
+        if os.readlink(path) != data.get('dest', ''):
+            res.add_error(path, 'link')
+
+    # Check directories are listed as directory
+    elif os.path.isdir(path):
+        if data['type'] != 'dir':
+            res.add_error(path, 'type')
+
+    else:
+        # Check file contents against hash and listed as file
+        # Check mtime and size as well
+        if stat.st_size != data['size']:
+            res.add_error(path, 'size')
+        if stat.st_mtime != data['time']:
+            res.add_error(path, 'mtime')
+        if data['type'] != 'file':
+            res.add_error(path, 'type')
+        if compute_hash(path) != data.get('hash', ''):
+            res.add_error(path, 'hash')
+
+    return res
+
+
+def check_file_manifest(file):
+    dirname = os.path.dirname(file)
+
+    results = VerificationResults()
+    while spack.store.layout.metadata_dir not in os.listdir(dirname):
+        if dirname == os.path.sep:
+            results.add_error(file, 'not owned by any package')
+            return results
+        dirname = os.path.dirname(dirname)
+
+    manifest_file = os.path.join(dirname,
+                                 spack.store.layout.metadata_dir,
+                                 spack.store.layout.manifest_file_name)
+
+    if not os.path.exists(manifest_file):
+        results.add_error(file, "manifest missing")
+        return results
+
+    try:
+        with open(manifest_file, 'r') as f:
+            manifest = json.load(f)
+    except Exception:
+        results.add_error(file, "manifest corrupted")
+        return results
+
+    if file in manifest:
+        results += check_entry(file, manifest[file])
+    else:
+        results.add_error(file, 'not owned by any package')
+    return results
+
+
+def check_spec_manifest(spec):
+    prefix = spec.prefix
+
+    results = VerificationResults()
+    manifest_file = os.path.join(prefix,
+                                 spack.store.layout.metadata_dir,
+                                 spack.store.layout.manifest_file_name)
+
+    if not os.path.exists(manifest_file):
+        results.add_error(prefix, "manifest missing")
+        return results
+
+    try:
+        with open(manifest_file, 'r') as f:
```

Comment:
I don't want a corrupted manifest file in one package to prevent us from verifying another package. So I'd rather record there was a problem and move on so I can print them all out later.

Reply:
I've factored out the two places it occurred, but check it again I'm not sure it was an improvement.

Reply:
This code is not a good candidate to refactor. It's 5 lines that require 3 lines to replace when you refactor, it only occurs twice, and it's tightly coupled with the variables in the code around it.

It requires 3 lines to replace when refactored because the calling code has to return based on whether an error occurred. Function call, conditional, return statement.

It requires returning a tuple because we need both the manifest we want and the error results if we generated any

And it requires passing in an unrelated filename, because we need to know what file to associate the error with.

I've reverted my attempts at refactoring this.

Reply:
Fair enough it is borderline for sure. My original concern was to encapsulate it in a function because if you handle it differently in one spot I assumed you'd want to adjust the others (so even if the total # of lines remained the same IMO it could still be an improvement) but again in this case admittedly borderline (and so I agree about skipping it).